app/domain/operations/people/match.rb

```diff
@@ -0,0 +1,156 @@
+      def call(params)
+        yield validate(params)
+        yield fetch_config_items
+        # query_params = yield query_builder(params)
+        person_records = yield match_person
+        yield build_result(person_records)
+      end
+
+      private
+
+      def validate_dob(dob)
+        if dob.nil?
+          false
+        else
+          dob.to_date.present?
+        end
+      end
+
+      # checks for any special characters, remove unnecessary special char and set the attrs as per values
+      # @return success without any return value
+      # @return failure for invalid params
+      def validate(params)
+        params.symbolize_keys!
+        result = params.all? do |k, v|
+          case k
+          when :dob
+            @has_dob = validate_dob(v)
+            true
+          when :ssn
+            @has_ssn = v&.delete('^0-9').present?
+            true
+          else
+            v.to_s.scan(SPECIAL_CHAR).blank?
+          end
+        end
+
+        if result
+          @first_name = params[:first_name]
+          @last_name = params[:last_name]
+          @dob = params[:dob].to_date
+          @ssn = params[:ssn]&.delete('^0-9')
+          Success("")
+        else
+          Failure("invalid params")
+        end
+      end
+
+      def fetch_config_items
+        if EnrollRegistry[:person_match_policy].enabled?
+          @keys_with_ssn = EnrollRegistry[:person_match_policy].settings(:ssn_present).item.map(&:to_sym)
+          @keys_with_dob = EnrollRegistry[:person_match_policy].settings(:dob_present).item.map(&:to_sym)
+          Success("")
+        else
+          Failure("person_match_policy is disabled")
+        end
+      end
+
+      # This method queries db as per the availability of the values
+      def match_person
+        result = if ssn.present?
+                   records = query({:dob => dob,
+                                    :encrypted_ssn => Person.encrypt_ssn(ssn)}).success
+                   records.count == 0 ? match_dob : [:ssn_dob, records]
+                 elsif dob.present?
+                   match_dob
+                 else
+                   [:name, query({
+                                   :last_name => /^#{last_name}$/i,
+                                   :first_name => /^#{first_name}$/i
+                                 }).success]
+                 end
+
+        Success(result)
+      end
+
+      def match_dob
+        [:name_dob, query({:dob => dob,
+                           :last_name => /^#{last_name}$/i,
+                           :first_name => /^#{first_name}$/i}).success]
+      end
+
+      # This method returns array of 3
+      # arr[0] is query criteria used to fetch the record
+      # arr[1] records in mongo criteria
+      # arr[2] error if any
+      #
+      # @example
+      #   build_result([initial_query_criteria, records])
+      #   # => Object
+      #
+      # @return [query_criteria, records, error]
+      def build_result(person_records)
```

Comment:
this method looks complex...do success or failure at the end of the method. 

you can assign query_options and return Success(query_options) at the end.